lib/readability/published_at_finder.ex

```diff
@@ -0,0 +1,88 @@
+defmodule Readability.PublishedAtFinder do
+  @moduledoc """
+  Extract the published at.
+  """
+
+  @type html_tree :: tuple | list
+
+  @strategies [:meta_tag, :time_element, :data_attribute, :text]
+
+  @doc """
+  Extract the published at.
+  """
+  @spec find(html_tree) :: %DateTime{} | %Date{} | nil
+  def find(html_tree) do
+    value =
+      Enum.find_value(@strategies, fn strategy ->
+        strategy(strategy, html_tree)
+      end)
+
+    if value do
+      parse(value)
+    end
+  end
+
+  defp strategy(:meta_tag, html_tree) do
+    selector = "meta[property='article:published_time'], meta[property='article:published']"
+
+    html_tree
+    |> Floki.find(selector)
+    |> Enum.flat_map(&Floki.attribute(&1, "content"))
+    |> Enum.map(&String.trim/1)
+    |> List.first()
+  end
+
+  defp strategy(:time_element, html_tree) do
+    html_tree
+    |> Floki.find("time")
+    |> Enum.flat_map(&Floki.attribute(&1, "datetime"))
+    |> Enum.map(&String.trim/1)
+    |> List.first()
+  end
+
+  defp strategy(:data_attribute, html_tree) do
+    html_tree
+    |> Floki.find("[data-datetime]")
+    |> Enum.flat_map(&Floki.attribute(&1, "data-datetime"))
+    |> Enum.map(&String.trim/1)
+    |> List.first()
+  end
+
+  defp strategy(:text, html_tree) do
+    html_tree
+    |> Floki.find("div, p, span")
```

Comment:
Isn't it too broad? `Floki.text` has "deep: true" by default, so what you're doing here is a lot. 

Eg. for 

```
<div>
  <p>Hello world</p>
  <div>
    <div>
       <span>Testing</span>
    </div>
  <div>
```

You're trying to parse dates 5 times, checking the same text multiple times. You can imagine how inefficient it can get with some deeply-nested HTMLs. Not to mention with that approach any valid date is considered published at, which IMO seems a bit too much? 

Eg in redability.js published_at seems to be extracted only from meta tags: https://github.com/mozilla/readability/pull/813/files.

Thoughts?

Reply:
Yeah, I agree it's not ideal. There's just so many that don't include it in meta tags or `<time>` elements, eg. the elixir example in the fixtures.

I did consider making it optional by allowing the consumer to optionally include the `date_time_parser` package. What do you think? I'm happy to remove it if you'd rather not have it at all though!

Reply:
@philipbrown I think I'd either remove it or make disabled by default? 

Or maybe we can make it use `Floki.text(doc, deep: false)` by default? 

Sth like `opts: [find_in_text: deep]`. 

Also. Wonder how DateTimeParser works exaclty. If there is a valid date inside the text, will it find it? 

If yes, maybe we could take the nested text from the outer-most div and run parser on it?

Reply:
Yeah, I'm just going to remove it. It's more of a nice to have, and it's hard to be confident that it's going to work consistently so I think it's better to just not have it.